Boost: Add cleanup for expired concatenated file paths

dilirity opened the pull request:
Fixes HOG-189

## Proposed changes:

* Set expiry of concatenated options to 1 week (was previously 1 year);
* Add cleanup for concatenated options that have expired.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

HOG-189

## Does this pull request change what data or activity we track or use?

no

## Testing instructions:

**Pre-requisites**

- Add this snippet to your site, to create some concatenated paths. Two of the file paths have expired so they should get deleted after cleanup runs:

```php
add_action( 'template_redirect', function () {
	if ( ! isset( $_GET['jb-test-concat-cleanup'] ) ) {
		return;
	}

	$option_prefix = 'jb_transient_concat_paths_';

	for ( $i = 1; $i <= 6; $i++ ) {
		$option_name = $option_prefix . $i;
		$expire = time();

		// Make options 3+ (inclusive), expire later.
		if ( $i >= 3 ) {
			$expire += MONTH_IN_SECONDS;
		}

		$option_value = array(
			'expire' => $expire,
			'data' => array(
				'paths' => array(
					'script-path-1.js',
					'script-path-2.js',
				),
			),
		);

		update_option( $option_name, $option_value );
	}
} );
```
- Open your home page with the get param `jb-test-concat-cleanup` to trigger the script;
- Check the database for options with name LIKE `jb_transient_c` (or run this SQL query `SELECT * FROM `wp_options` WHERE `option_name` LIKE '%jb_transient_c%'`. There should be 6 file path Boost transients:

<img width="859" alt="CleanShot 2025-06-30 at 14 26 00@2x" src="https://github.com/user-attachments/assets/1e68c552-3d59-47e6-9463-b232608dd7a7" />

- [WP Crontrol plugin](https://wordpress.org/plugins/wp-crontrol/) installed - it would make debugging schedules easier.

### Test that cleanup works

- Setup Boost (free) from this branch;
- Open `WP Admin -> Tools -> Cron Events` and search for `clean` to make sure that `jetpack_boost_minify_cron_cleanup_concat_paths` and `jetpack_boost_minify_cron_cleanup_concat_paths_followup` are not in the list;
- Enable Concatenate JS (or Concatenate CSS);
- Check `Tools -> Cron Events` again and there should be an event called `jetpack_boost_minify_cron_cleanup_concat_paths`;
- Hover over the event, and click `Run Now` (to force schedule to run cleanup);
- Since WP schedules need a visit to the website, open the home page after you run the script (this will simulate a user/bot visit to the website);
- Now check the entries in the DB with the query from the **Pre-Requisites** section and entries 1 and 2 should be gone (you might see other entries, but that's probably because the home page visit concatenated some assets):

<img width="928" alt="CleanShot 2025-06-30 at 14 29 49@2x" src="https://github.com/user-attachments/assets/608f7317-208e-4ba7-aef9-c2c181f223ea" />

### Test that cleanup schedules are removed

- Disable both `Concatenate JS` and `Concatenate CSS`;
- Visit `Tools -> Cron Events` and search for `clean` to make sure that `jetpack_boost_minify_cron_cleanup_concat_paths` and `jetpack_boost_minify_cron_cleanup_concat_paths_followup` are not in the list;
- Uninstalling Boost with either `Concatenate JS` or `Concatenate CSS` active should remove the schedules up as well.

dilirity committed:
Add cleanup for expired concatenated file paths

dilirity committed:
Add changelog

dilirity committed:
Update file paths expiry to 1 week

dilirity committed:
Update file paths expiry to 1 month

dilirity committed:
Fix cleanup removing week old files instead of 1 month old files

dilirity committed:
Merge branch 'trunk' into add/boost/minify-cleanup-HOG-189

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/class-cleanup-stored-paths.php:
```diff
@@ -0,0 +1,127 @@
+	 * @var string
+	 */
+	private $last_processed_option_key = 'jb_transient_cleanup_concat_paths_last_processed_option_id';
+
```
Per offline discussion, let's prefix this with `jetpack_boost_` instead of `jb_transient_`

```suggestion
	private $last_processed_option_key = 'jetpack_boost_cleanup_concat_paths_last_processed_option_id';
```

dilirity replied:
I didn't check the diff and just did a commit :D

3a42244225fe455fe658dc084205268003300997

Make sure you select the lines you want to change, so it removes the old one.

LiamSarsfield replied:
Oh interesting! I used the VS Code GitHub PR plugin for the suggestion so I'm blaming that :D Will double check this in the future.

dilirity replied:
No worries! GH actions caught it.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/class-cleanup-stored-paths.php:
```diff
@@ -0,0 +1,127 @@
+				continue;
+			}
+
+			if ( $value['expire'] <= $expire_timestamp ) {
```
Per offline discussion, let's just use `time()` here instead of the `$expire_timestamp`, as the expire value is already the time which the path needs to be deleted.
(I would use a suggestion here, but we also should remove the `$expire_timestamp` variable).

dilirity replied:
Thanks for catching this!

Fixed in 2fde1e4e6b94e44b08db2c75e911f3f863802f24.

dilirity committed:
Update option name

It makes more sense for the option name to be a generic jetpack boost option, instead of a transient (which might be confusing).

Co-authored-by: Liam Sarsfield <43409125+LiamSarsfield@users.noreply.github.com>

dilirity committed:
Fix deleting entries that haven't expired yet

dilirity committed:
Remove duplicate variable declaration

dilirity committed:
Fix adding multiple schedules

dilirity committed:
Avoid scheduling a follow up clean up if there are no more entries

dilirity commented:
@LiamSarsfield thanks for the awesome idea with the offset!

I've implemented the change in 9598f32.

LiamSarsfield reviewed (approved):
LGTM! Works as expected, also tested it with 150+ transients, and 100 expired ones. Works like a charm

kraftbj committed:
Merge branch 'trunk' into add/boost/minify-cleanup-HOG-189

kraftbj commented on projects/plugins/boost/app/lib/minify/class-cleanup-stored-paths.php:
```diff
@@ -0,0 +1,149 @@
+			}
+
+			if ( $value['expire'] <= time() ) {
+				delete_option( $option['option_name'] );
```
should we do some type of clean up for the existing year-long transients? If the value is either less than the current time (truly expired) or if the value is more than... 45 days in the future?

With sufficient inline documentation to explain why we care about that...

dilirity replied:
We considered cleaning up transients that are year-long. Ultimately, we decided against cleaning them up, as removing them might do more harm than good.

The harm _might_ come from cached pages. If we were to delete such transients, cached pages referencing the deleted hashes will break (as we'll also be [cleaning up files](https://github.com/Automattic/jetpack/pull/44137) belonging to those hashes), because there's no data to reconstruct the file from.

Since such options aren't auto-loaded, they don't impact the website that much. They would get cleaned eventually. We'll also be updating the Boost uninstall to use a direct DB query, which should be way faster than the regular delete_option (which was the original problem).
